napari/components/_viewer/viewer.py

```diff
@@ -210,18 +221,59 @@ def _update_layer_selection(self, event):
         # iteration goes backwards to find top most selected layer if any
         for layer in self.layers[::-1]:
             if layer.selected:
-                self._qt.control_panel.display(layer)
+                self._qtviewer.control_panel.display(layer)
                 self.status = layer.status
                 self.help = layer.help
                 self.cursor = layer.cursor
                 self.interactive = layer.interactive
                 self._top = layer
                 break
         else:
-            self._qt.control_panel.display(None)
+            self._qtviewer.control_panel.display(None)
             self.status = 'Ready'
             self.help = ''
             self.cursor = 'standard'
             self.interactive = True
             self._top = None
-        self._canvas.native.setFocus()
+        self._qtviewer.canvas.native.setFocus()
+
+    def _on_layers_change(self, event):
+        self.dims.set_all_ranges(self._calc_layers_ranges())
+
+    def _calc_layers_ranges(self):
+        """Calculates the range along each axis from all present layers.
+        """
+
+        ndims = self._calc_layers_num_dims()
+        ranges = [(inf, -inf, inf)]*ndims
+
+        for layer in self.layers:
+            layer_range = layer.range[::-1]
+            ranges = [(min(a, b), max(c, d), min(e, f)) for
+                      (a, c, e), (b, d, f) in zip_longest(ranges, layer_range,
```

Comment:
@jni please review - I had to use `zip_longest` to make sure we were getting all the dims info across layers